urns/urns.go

```diff
@@ -84,36 +85,43 @@ type URN string
 
 // NewTelURNForCountry returns a URN for the passed in telephone number and country code ("US")
 func NewTelURNForCountry(number string, country string) (URN, error) {
-	return NewURNFromParts(TelScheme, normalizeNumber(number, country), "")
+	return NewValidatedURNFromParts(TelScheme, normalizeNumber(number, country), "", "")
 }
 
 // NewTelegramURN returns a URN for the passed in telegram identifier
 func NewTelegramURN(identifier int64, display string) (URN, error) {
-	return NewURNFromParts(TelegramScheme, strconv.FormatInt(identifier, 10), display)
+	return NewValidatedURNFromParts(TelegramScheme, strconv.FormatInt(identifier, 10), "", display)
 }
 
 // NewWhatsAppURN returns a URN for the passed in whatsapp identifier
 func NewWhatsAppURN(identifier string) (URN, error) {
-	return NewURNFromParts(WhatsAppScheme, identifier, "")
+	return NewValidatedURNFromParts(WhatsAppScheme, identifier, "", "")
 }
 
 // NewFirebaseURN returns a URN for the passed in firebase identifier
 func NewFirebaseURN(identifier string) (URN, error) {
-	return NewURNFromParts(FCMScheme, identifier, "")
+	return NewValidatedURNFromParts(FCMScheme, identifier, "", "")
 }
 
 // NewFacebookURN returns a URN for the passed in facebook identifier
 func NewFacebookURN(identifier string) (URN, error) {
-	return NewURNFromParts(FacebookScheme, identifier, "")
+	return NewValidatedURNFromParts(FacebookScheme, identifier, "", "")
 }
 
-// NewURNFromParts returns a new URN for the given scheme, path and display
-func NewURNFromParts(scheme string, path string, display string) (URN, error) {
-	urnString := fmt.Sprintf("%s:%s", scheme, path)
-	if display != "" {
-		urnString = fmt.Sprintf("%s#%s", urnString, display)
+// NewURNFromParts returns a new URN for the given scheme, path, query and display
+func NewURNFromParts(scheme string, path string, query string, display string) URN {
```

Comment:
we do a lot of splitting and re-forming URNs, and we don't always want to do validation

Reply:
Can this be private though? I don't think validation is the worst thing in the world from a public interface point of view

Reply:
Also like the idea of if you have a URN you know it is valid no matter what.